test/lit/passes/code-folding.wast

```diff
@@ -481,10 +591,243 @@
     (unreachable)
   )
  )
-)
 
-(module
- ;; CHECK:      (type $0 (func))
+ ;; CHECK:      (func $br-with-value (type $2) (result i32)
+ ;; CHECK-NEXT:  (block $l
+ ;; CHECK-NEXT:   (block
+ ;; CHECK-NEXT:    (br $l)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (block
+ ;; CHECK-NEXT:    (br $l)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (i32.const 1)
+ ;; CHECK-NEXT: )
+ (func $br-with-value (result i32)
+  (block $l (result i32)
+   (block
+    (br $l
+     (i32.const 1)
+    )
+   )
+   (block
+    (br $l
+     (i32.const 1)
+    )
+   )
+   (i32.const 1)
+  )
+ )
+
+ ;; CHECK:      (func $br-and-fallthrough-with-value (type $2) (result i32)
+ ;; CHECK-NEXT:  (block $l
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (block (result i32)
+ ;; CHECK-NEXT:     (br $l)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (block (result i32)
+ ;; CHECK-NEXT:     (br $l)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (i32.const 1)
+ ;; CHECK-NEXT: )
+ (func $br-and-fallthrough-with-value (result i32)
+  (block $l (result i32)
+   (drop
+    (block (result i32)
+     (br $l
+      (i32.const 1)
+     )
+    )
+   )
+   (drop
+    (block (result i32)
+     (br $l
+      (i32.const 1)
+     )
+    )
+   )
+   (i32.const 1)
+  )
+ )
+
+ ;; CHECK:      (func $br-with-value-and-more (type $2) (result i32)
+ ;; CHECK-NEXT:  (block $l
+ ;; CHECK-NEXT:   (block
+ ;; CHECK-NEXT:    (br $l)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (block
+ ;; CHECK-NEXT:    (nop)
+ ;; CHECK-NEXT:    (br $l)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (nop)
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (nop)
+ ;; CHECK-NEXT:  (nop)
+ ;; CHECK-NEXT:  (i32.const 1)
```

Comment:
This doesn't look right. Imagine the inner blocks are ifs, and that we get false in both conditions, so we enter neither. Then before, all we execute is one nop and then return the constant. But after, we execute two nops. (If the nops were say calls, then we'd have possibly different effects as a result.)

Reply:
The correctness here depends on the fact that the inner blocks are unreachable. This means that the fallthrough tail cannot possibly be executed, so it is not considered when merging tails. This test therefore moves two nops instead of just one. The very next test is the same except that the blocks are not unreachable, so it considers the fallthrough tail and only moves one nop.

Reply:
Note that this special casing for blocks containing unreachable instructions existed before this PR, but was not tested, at least not in lit tests.

Reply:
I see, thanks!